javalib/src/main/scala/java/util/ArrayList.scala

```diff
@@ -14,80 +14,187 @@ package java.util
 
 import java.lang.Cloneable
 import java.lang.Utils._
+import java.util.ScalaOps._
 
 import scala.scalajs._
+import scala.scalajs.LinkingInfo.isWebAssembly
 
-class ArrayList[E] private (private[ArrayList] val inner: js.Array[E])
+class ArrayList[E] private (innerInit: AnyRef, private var _size: Int)
     extends AbstractList[E] with RandomAccess with Cloneable with Serializable {
   self =>
 
+  /* This class has two different implementations for handling the
+   * internal data storage, depending on whether we are on Wasm or JS.
+   * On JS, we utilize `js.Array`. On Wasm, for performance reasons,
+   * we avoid JS interop and use a scala.Array.
+   * The `_size` field (unused in JS) keeps track of the effective size
+   * of the underlying Array for the Wasm implementation.
+   */
+
+  private val innerJS =
+    if (isWebAssembly) null
+    else innerInit.asInstanceOf[js.Array[E]]
+
+  private var innerWasm =
+    if (!isWebAssembly) null
+    else innerInit.asInstanceOf[Array[Any]]
```

Comment:
I suggest creating this array as an `Array[AnyRef]`. They erase to the same thing, but `Array[AnyRef]` will allow to use the methods of `Arrays`. It's common practice even inside the Scala standard library to use an `Array[AnyRef]` when we need a performant "array of anything".

Note that it is always valid to cast to `AnyRef`, even for values that "are `AnyVal`s". `AnyVal` is a Scala compiler fiction. It erases to `AnyRef` as well. Casting to `AnyRef` is a no-op in the IR/bytecode.

Reply:
> Note that it is always valid to cast to AnyRef, even for values that "are AnyVals". AnyVal is a Scala compiler fiction. It erases to AnyRef as well. Casting to AnyRef is a no-op in the IR/bytecode.

Interesting, I'll play around with it, but for now we cast it to `Array[AnyRef]` and can use the `Arrays` functions. Thanks!